api/thirdpartyresource/helpers.go

```diff
@@ -0,0 +1,133 @@
+package thirdpartyresource
+
+import (
+	"fmt"
+	"github.com/docker/swarmkit/api"
+	"strings"
+)
+
+// NewThirdPartyResourceError creates a new error for Third Party Resources
+func NewThirdPartyResourceError(format string, args ...interface{}) error {
+	return fmt.Errorf("Could not parse TPR: "+format, args...)
+}
+
+// ValidateTaskThirdPartyResources validates that the task only uses integers
+// for Third Party Resources
+func ValidateTaskThirdPartyResources(resources *api.Resources) error {
+	for k, v := range resources.ThirdParty {
+		switch v.Resource.(type) {
+		case *api.ThirdPartyResource_Integer:
+			continue
+		case *api.ThirdPartyResource_Set:
+			return NewThirdPartyResourceError("Invalid Argument for resource %s", k)
+		}
+	}
+
+	return nil
+}
+
+func getTPRInteger(res *api.ThirdPartyResource) (*api.ThirdPartyResource_Integer, error) {
+	if _, ok := res.Resource.(*api.ThirdPartyResource_Integer); !ok {
+		return nil, fmt.Errorf("Task should only hold Integer type Third Party Resources")
+	}
+
+	return res.Resource.(*api.ThirdPartyResource_Integer), nil
+}
+
+// HasEnoughThirdPartyResources returns true if node can satisfy task TPR's request
+func HasEnoughThirdPartyResources(nodeRes *api.ThirdPartyResource, taskRes *api.ThirdPartyResource) (bool, error) {
+	t, err := getTPRInteger(taskRes)
+	if err != nil {
+		return false, err
+	}
+
+	if nodeRes == nil {
+		return false, nil
+	}
+
+	switch nr := nodeRes.Resource.(type) {
+	case *api.ThirdPartyResource_Integer:
+		if t.Integer.Val > nr.Integer.Val {
+			return false, nil
+		}
+	case *api.ThirdPartyResource_Set:
+		if t.Integer.Val > uint64(len(nr.Set.Val)) {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+// ClaimThirdPartyResource assign TPRs to the task by taking them from the node's TPR
+// and storing them in the taken store
+func ClaimThirdPartyResource(nodeRes *api.ThirdPartyResource, taskRes *api.ThirdPartyResource, taken *api.ThirdPartyResource) error {
+	t, err := getTPRInteger(taskRes)
+	if err != nil {
+		return err
+	}
+
+	switch nr := nodeRes.Resource.(type) {
+	case *api.ThirdPartyResource_Integer:
+		nr.Integer.Val -= t.Integer.Val
+		taken.Resource = &api.ThirdPartyResource_Integer{
+			Integer: &api.Integer{Val: t.Integer.Val}}
```

Comment:
Cosmetic nit: I usually see this written as

```go
		taken.Resource = &api.ThirdPartyResource_Integer{
			Integer: &api.Integer{Val: t.Integer.Val},
		}
```

It just seems a little more "balanced" that way.